codex/stores/blockstore.nim

```diff
@@ -27,23 +29,31 @@ type
   BlockType* {.pure.} = enum
     Manifest, Block, Both
 
-  GetNext* = proc(): Future[?Cid] {.upraises: [], gcsafe, closure.}
+  BlockStore* = ref object of RootObj
 
-  BlocksIter* = ref object
-    finished*: bool
-    next*: GetNext
+method getBlock*(self: BlockStore, cid: Cid): Future[?!Block] {.base.} =
+  ## Get a block from the blockstore
+  ##
 
-  BlockStore* = ref object of RootObj
+  raiseAssert("getBlock by cid not implemented!")
 
-iterator items*(self: BlocksIter): Future[?Cid] =
-  while not self.finished:
-    yield self.next()
+method getBlock*(self: BlockStore, treeCid: Cid, index: Natural): Future[?!Block] {.base.} =
```

Comment:
Ad1. But backwards index is a different access pattern and hence a different parameter type. This is what we have in `AsyncHeapQueue`
```nim
proc `[]`*[T](heap: AsyncHeapQueue[T], i: Natural) : T {.inline.} =
  ## Access the i-th element of ``heap`` by order from first to last.
  ## ``heap[0]`` is the first element, ``heap[^1]`` is the last element.
  heap.queue[i]

proc `[]`*[T](heap: AsyncHeapQueue[T], i: BackwardsIndex) : T {.inline.} =
  ## Access the i-th element of ``heap`` by order from first to last.
  ## ``heap[0]`` is the first element, ``heap[^1]`` is the last element.
  heap.queue[len(heap.queue) - int(i)]
```

Ad2. There's a method to get block by address (line 46), but in some cases we don't have an address object in the scope, instead we have separately `treeCid` and `index` variables. We could allocate the memory and create the `BlockAddress` object, but: 1. that seem a bit wasteful and 2. it's an additional inconvenience to the client.

Reply:
> Ad1. But backwards index is a different access pattern and hence a different parameter type. This is what we have in AsyncHeapQueue

OK, not a big deal - we can't use `[]` operator with async and we'll unlikely to use a `BackwardsIndex` either way.

> Ad2. There's a method to get block by address (line 46), but in some cases we don't have an address object in the scope, instead we have separately treeCid and index variables. We could allocate the memory and create the BlockAddress object, but: 1. that seem a bit wasteful and 2. it's an additional inconvenience to the client.

Hmm, that seems like a problem of not using `BlockAddress` consistently rather than anything else?

Reply:
>Hmm, that seems like a problem of not using BlockAddress consistently rather than anything else?

So in couple of places we have a following pattern:
```nim
for i in 0..<manifest.blocksCount:
   await getBlock(manifest.treeCid, i)
```

Would you like it to be replaced by following code?
```nim
for i in 0..<manifest.blocksCount:
   let address = BlockAddress.init(manifest.treeCid, i)
   await getBlock(address)
```

If so I could get rid of the method `getBlock(treeCid: Cid, index: Natural)`, but in my opinion the 2nd code is simply worse because it's:
- less efficient (new object created on each iteration)
- more verbose than necessary

Reply:
Yeah, I understand the reasoning, but IMO having two slightly distinct interfaces will just make the codebase less consistent - IMO, consistency trumps everything else.

We could add an initializer from a tuple for example 

```nim
init(_: type (Cid, Natural)): BlockAddress = ...
```

The point is that we stick with one convention and follow that, otherwise we'll end up with (slightly) distinct looking flows, which will make the codebase harder to read and maintain.

Reply:
Alright. I understand the reasoning and I agree that unification has certain benefits (if at some point we will remove fetching individual blocks by just `cid`, we will have less to refactor).

Also there's more to unify if we go this path:

```nim
method ensureExpiry*(self: BlockStore, cid: Cid, expiry: SecondsSince1970)

method delBlock*(self: BlockStore, cid: Cid)
method delBlock*(self: BlockStore, treeCid: Cid, index: Natural)

method hasBlock*(self: BlockStore, cid: Cid)
method hasBlock*(self: BlockStore, tree: Cid, index: Natural)

proc contains*(self: BlockStore, blk: Cid)
proc contains*(self: BlockStore, address: BlockAddress)
````

I could do this as a part of this PR or in a separate one. Wdyt?